### PR TITLE
Alembic migration

### PR DESCRIPTION
**Pull request by sc68cal**

### CODE REVIEW

**Comment by sc68cal** [comment]
Current issues

- [ ] Update the integration script to use the alembic migration

**Comment by sc68cal** [comment]
Need to fix up the commit message to have it fix #35

File: kostyor/db/migrations/alembic/env.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+# Interpret the config file for Python logging.
+# This line sets up loggers basically.
+fileConfig(config.config_file_name)
+
+# add your model's MetaData object here
+# for 'autogenerate' support
+# from myapp import mymodel
+# target_metadata = mymodel.Base.metadata
+target_metadata = None
```
Comment (ikalnitsky) [inline kostyor/db/migrations/alembic/env.py]: I think we should point it to `models.Base.metadata` in order to support `autogenerate` feature. Could you please specify it?

File: kostyor/db/migrations/alembic/env.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+
+def run_migrations_online():
+    """Run migrations in 'online' mode.
+
+    In this scenario we need to create an Engine
+    and associate a connection with the context.
+
+    """
+    connectable = engine_from_config(
```
Comment (ikalnitsky) [inline kostyor/db/migrations/alembic/env.py]: I think in order to have a single source of truth it makes sense to get connection not from Alembic config, but by direct import of engine from `db/api.py`. In that case, in order to fix #37 we will need to just configure engine properly in `db/api.py`.

What do you think?

File: kostyor/db/migrations/alembic/versions/2bedb47d68e6_initial_schema.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+    op.create_table(
+        'services',
+        sa.Column('name', sa.String(255)),
+        sa.Column('id', sa.String(36), nullable=False),
+        sa.Column('host_id', sa.String(36), nullable=False),
+        sa.Column('version', sa.Enum(*VERSIONS)),
+        sa.ForeignKeyConstraint(['host_id'], ['hosts.id']),
+        sa.PrimaryKeyConstraint('id')
+
```
Comment (ikalnitsky) [inline kostyor/db/migrations/alembic/versions/2bedb47d68e6_initial_schema.py]: Could you please remove this empty line?

File: kostyor/db/migrations/alembic/versions/2bedb47d68e6_initial_schema.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    )
+
+    op.create_table(
+        'upgrade_tasks',
+        sa.Column('id', sa.String(36), nullable=False),
+        sa.Column('upgrade_start_time', sa.DateTime),
```
Comment (ikalnitsky) [inline kostyor/db/migrations/alembic/versions/2bedb47d68e6_initial_schema.py]: Proposal: since the table is called `upgrade_tasks` what do you think about renaming this column into just `start_time`?

File: kostyor/db/migrations/alembic/env.py (hunk 2 above)
Comment (sc68cal) [inline kostyor/db/migrations/alembic/env.py]: yes that is the point of #37 - which I state is blocking #35

File: kostyor/db/migrations/alembic/env.py (hunk 1 above)
Comment (sc68cal) [inline kostyor/db/migrations/alembic/env.py]: No, I don't believe in using autogeneration - I've used it a couple times in the past and was not happy with the DDL that was generated.

Comment (ikalnitsky) [inline kostyor/db/migrations/alembic/env.py]: Could you please elaborate on what exactly was wrong? You've got to have the same DDL anyway.

Besides, manual writing mean it's much easier to get out of sync database schema produced by ORM and database schema produced by Alembic. In that case having a special test to check that there's no difference is a must even on early stages.

File: kostyor/db/migrations/alembic/versions/2bedb47d68e6_initial_schema.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+revision = '2bedb47d68e6'
+down_revision = None
+branch_labels = None
+depends_on = None
+
+STATUSES = ['READY FOR UPGRADE', 'UPGRADE IN PROGRESS', 'UPGRADE PAUSED',
+            'UPGRADE ERROR', 'UPGRADE CANCELLED', 'NOT READY FOR UPGRADE',
+            'UPGRADE ROLLBACK IN PROGRESS']
+
+VERSIONS = ['liberty', 'mitaka', 'newton', 'unknown']
```
Comment (dstepanenko) [inline kostyor/db/migrations/alembic/versions/2bedb47d68e6_initial_schema.py]: let's sync the order with open api spec https://github.com/sc68cal/Kostyor/pull/36/commits/6a62e365d76100d1d041e06196d815f0702e2327#diff-3550eb108976e83d6ab7269c59348b7fR155

File: kostyor/db/migrations/alembic/versions/2bedb47d68e6_initial_schema.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+import sqlalchemy as sa
+
+
+# revision identifiers, used by Alembic.
+revision = '2bedb47d68e6'
+down_revision = None
+branch_labels = None
+depends_on = None
+
+STATUSES = ['READY FOR UPGRADE', 'UPGRADE IN PROGRESS', 'UPGRADE PAUSED',
```
Comment (dstepanenko) [inline kostyor/db/migrations/alembic/versions/2bedb47d68e6_initial_schema.py]: open API spec for Kostyor https://github.com/sc68cal/Kostyor/pull/36/commits/6a62e365d76100d1d041e06196d815f0702e2327#diff-3550eb108976e83d6ab7269c59348b7fR162 contains  '_" instead of whitespace in the statuses. Let's sync it with spec or create an issue tfor changing open API spec in the future accroding to this. Also both statuses order and text of the statuses are different from the one used in open  API spec

File: kostyor/db/migrations/alembic/versions/2bedb47d68e6_initial_schema.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+    )
+
+    op.create_table(
+        'upgrade_tasks',
+        sa.Column('id', sa.String(36), nullable=False),
+        sa.Column('upgrade_start_time', sa.DateTime),
+        sa.Column('upgrade_end_time', sa.DateTime),
+        sa.Column('cluster_id', sa.String(36), nullable=False),
+        sa.Column('from_version', sa.Enum(*VERSIONS)),
```
Comment (dstepanenko) [inline kostyor/db/migrations/alembic/versions/2bedb47d68e6_initial_schema.py]: Just an idea - shouldn't we create a constraint to guarantee that from_version will be less than to_version?

File: kostyor/db/migrations/alembic/versions/2bedb47d68e6_initial_schema.py (hunk 6 above)
Comment (ikalnitsky) [inline kostyor/db/migrations/alembic/versions/2bedb47d68e6_initial_schema.py]: I think fixing constants should be done in separate patch, not in this one. We use space-based constants in the code anyway.

File: kostyor/db/migrations/alembic/versions/2bedb47d68e6_initial_schema.py (hunk 7 above)
Comment (ikalnitsky) [inline kostyor/db/migrations/alembic/versions/2bedb47d68e6_initial_schema.py]: How you're going to do this? It's a string. I think this sort of checks should be done on backend side during request validation.

File: kostyor/db/migrations/alembic/versions/2bedb47d68e6_initial_schema.py (hunk 4 above)
Comment (sc68cal) [inline kostyor/db/migrations/alembic/versions/2bedb47d68e6_initial_schema.py]: In a separate patch, with a migration, sure.

File: kostyor/db/migrations/alembic/versions/2bedb47d68e6_initial_schema.py (hunk 3 above)
Comment (sc68cal) [inline kostyor/db/migrations/alembic/versions/2bedb47d68e6_initial_schema.py]: Please, let's not start reviewing with an eye towards whitespace. C'mon.
